tests/mr/device/callback_mr_tests.cpp

```diff
@@ -0,0 +1,80 @@
+/*
+ * Copyright (c) 2021, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "../../byte_literals.hpp"
+#include "rmm/cuda_stream_view.hpp"
+#include "rmm/mr/device/device_memory_resource.hpp"
+
+#include <cstddef>
+#include <rmm/detail/error.hpp>
+#include <rmm/device_buffer.hpp>
+#include <rmm/mr/device/callback_memory_resource.hpp>
+
+#include <spdlog/fmt/ostr.h>
+
+#include <gtest/gtest.h>
+
+namespace rmm::test {
+namespace {
+
+TEST(CallbackTest, PassThroughTest)
+{
+  auto base_mr           = rmm::mr::get_current_device_resource();
+  auto allocate_callback = [&base_mr](std::size_t size, cuda_stream_view stream, void* arg) {
+    return base_mr->allocate(size, stream);
+  };
+  auto deallocate_callback = [&base_mr](
+                               void* ptr, std::size_t size, cuda_stream_view stream, void* arg) {
+    base_mr->deallocate(ptr, size, stream);
+  };
+  auto mr =
+    rmm::mr::callback_memory_resource(allocate_callback, deallocate_callback, nullptr, nullptr);
+  auto ptr = mr.allocate(10_MiB);
+  mr.deallocate(ptr, 10_MiB);
+}
```

Comment:
maybe verify `base_mr` used space changes by 10 megs?

Reply:
I don't think we can do that for any arbitrary `base_mr`. I've added a check that the returned `ptr` is not `nullptr`. Is that sufficient?

Reply:
The right answer is probably to use something like a mock'ed memory_resource where you just verify it gets called in the way you expect. 

Alternatively you could do something like with the logging resource tests where you wrap the `base_mr` in a logging resource and verify the generated log is what you expect based on the usage of the callback resource.

https://github.com/rapidsai/rmm/blob/36dd42db02fbda76b213c320e389c7c4a52510dd/tests/logger_tests.cpp#L155-L180

Reply:
This is pretty close to what the test below does, where maybe we can just get rid of this test?

Reply:
Ah I see but we don't test that the "context" arguments are handled correctly then.

Reply:
@jrhemstad I found that the `aligned_memory_resource` were already using a mock class like you described. I've pulled that out into a header of its own so I could reuse here - let  me know if you don't like that change.

Reply:
Nice, that looks great!